src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -891,9 +907,12 @@ private void NotifyInvoker(string functionName, Exception ex)
 
         private void OnFileChanged(object sender, FileSystemEventArgs e)
         {
-            string fileName = Path.GetFileName(e.Name);
+            string directory = GetRelativeDirectory(e.FullPath, ScriptConfig.RootScriptPath);
+            bool isWatchedDirectory = ScriptConfig.WatchDirectories.Contains(directory, StringComparer.OrdinalIgnoreCase);
 
-            if (((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
+            string fileName = Path.GetFileName(e.Name);
+            if (isWatchedDirectory ||
+                ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot))
```

Comment:
This conditional is a bit hard to read through and could use a comment. e.g. explicitely list in plain English the conditions that trigger a restart. I'd also yank a bunch of extra parenthesis which makes it look more complicated that it probably is :)